pkg/error_code/error_code.go

```diff
@@ -166,41 +265,58 @@ func (e CodedError) GetClientData() interface{} {
 // invalidInput gives the code InvalidInputCode
 type invalidInputErr struct{ CodedError }
 
-// NewInvalidInputErr creates an invalidInput with error code of InvalidInputCode
+// NewInvalidInputErr creates an invalidInput from an err
+// If the error is already an ErrorCode it will use that code
+// Otherwise it will use InvalidInputCode which gives HTTP 400
 func NewInvalidInputErr(err error) ErrorCode {
-	return invalidInputErr{CodedError{RegisteredCode: InvalidInputCode, Err: err}}
+	return invalidInputErr{NewCodedError(err, InvalidInputCode)}
 }
 
-var _ ErrorCode = (*invalidInputErr)(nil) // assert implements interface
+var _ ErrorCode = (*invalidInputErr)(nil)     // assert implements interface
+var _ HasClientData = (*invalidInputErr)(nil) // assert implements interface
 
 // internalError gives the code InvalidInputCode
 type internalErr struct{ CodedError }
 
-// GetHTTPCode returns 500
-func (e internalErr) GetHTTPCode() int {
-	return http.StatusInternalServerError
-}
-
-// NewInternalErr creates an internalError with error code InternalCode
+// NewInternalErr creates an internalError from an err
+// The code is always set to InternalCode which returns HTTP 500
 func NewInternalErr(err error) ErrorCode {
-	return internalErr{CodedError{RegisteredCode: InternalCode, Err: err}}
+	return internalErr{CodedError{GetCode: InternalCode, Err: err}}
```

Comment:
Why not use `NewCodedError` here?

Reply:
Its behavior is different to not allow the intention of sending a 50X to be overidden. I added some more comments and a better implementation that allows for descendants.